logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,215 @@
+    private static final String KEYSTORE_TYPE = "pkcs12";
+
+    private final char[] keyStorePass;
+    private final Path keyStorePath;
+    private final ProtectionParameter protectionParameter;
+    private final Lock readLock;
+    private final Lock writeLock;
+    private KeyStore keyStore;
+
+    /**
+     * Constructor - will create the keystore if it does not exist
+     *
+     * @param keyStorePath The full path to the java keystore
+     * @param keyStorePass The password to the keystore
+     * @throws SecretStoreException if errors occur while trying to create or access the keystore
+     */
+    public JavaKeyStore(Path keyStorePath, char[] keyStorePass) {
+        try {
+            this.keyStorePath = keyStorePath;
+            this.keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
+            this.keyStorePass = SecretStoreUtil.base64Encode(keyStorePass);
+            ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+            readLock = readWriteLock.readLock();
+            writeLock = readWriteLock.writeLock();
+            this.protectionParameter = new PasswordProtection(this.keyStorePass);
+            SecretIdentifier logstashMarker = new SecretIdentifier(LOGSTASH_MARKER);
+
+            try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                try {
+                    keyStore.load(is, this.keyStorePass);
+                } catch (IOException ioe) {
+                    if (ioe.getCause() instanceof UnrecoverableKeyException) {
+                        throw new SecretStoreException.AccessException(
+                                String.format("Can not access Java keystore at %s, check file permissions and the keystore password", keyStorePath.toAbsolutePath
+                                        ()), ioe);
+                    } else {
+                        throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a file at %s, but it is not a valid logstash keystore.",
+                                keyStorePath.toAbsolutePath().toString()), ioe);
+                    }
+                }
+                byte[] marker = retrieveSecret(logstashMarker);
+                if (marker == null) {
+                    throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a keystore at %s, but it is not a logstash keystore.",
+                            keyStorePath.toAbsolutePath().toString()));
+                }
+                LOGGER.debug("Using existing keystore at {}", keyStorePath.toAbsolutePath().toString());
+            } catch (NoSuchFileException noSuchFileException) {
+                LOGGER.info("Keystore not found at {}. Creating new keystore.", keyStorePath.toAbsolutePath().toString());
+                //create the keystore on disk with a default entry to identify this as a logstash keystore
+                try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
```

Comment:
Thinking about this more, I think this is not a secure way to create this file.

We want to create this file and set the permissions simultaneously. `Files.newOutputStream` doesn't let you do this. Instead, we should probably do this:

```java
import java.nio.file.Paths;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.PosixFilePermissions;
public class X {
  public static void main(String ... args) throws Exception {
    Path path = Paths.get(args[0]);
    Files.createFile(path, PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rw-------")));
    //OutputStream os = new FileOutputStream(path.toFile());
  }
}
```

Observing the syscalls on Linux, the file permissions are given to the `open(2)` call:

```
% rm -f /tmp/x; strace -fe trace=file java X /tmp/x |& grep /tmp/x
execve("/usr/bin/java", ["java", "X", "/tmp/x"], 0x7fff34f82710 /* 36 vars */) = 0
[pid 47040] open("/tmp/x", O_WRONLY|O_CREAT|O_EXCL, 0600) = 5
```

This ensures the file never exists without safe access control.

Reply:
Nice,  but not perfect. 

https://docs.oracle.com/javase/tutorial/essential/io/fileAttr.html#posix
> The code then attempts to create a new file with those permissions. Note that the umask also applies, so the new file might be more secure than the permissions that were requested.

...so if we are trying to set "rw-rw----", but the umask is for the directory is "rwxr-xr-x" , you end up with "rw-r-----".

A couple options here... but I propose the following

Set it in both places (at create and afterwards).
Remove the code that deletes and throws if it can not set it afterwards, since that would only now be expanding permissions.